### ISSUE TITLE
Data of user-defined dimensions and tensors of higher ranks

### ISSUE THREAD

**Issue by fsimonis** [precice/precice#2085]
**Please describe the problem you are trying to solve.**

preCICE currently supports tensors of rank 0 and 1, namely scalar and vectorial data.
Some applications need to communicate tensors of higher ranks or even data of custom dimensionality.

The current workaround is to use either pseudo meshes or multiple meshes to communicate the data via preCICE.

**Describe the solution you propose.**

Add `tensor` and/or `custom` data types. Names up for discussion as usual.

What we currently support:
```xml
<data:scalar name="S" />
<data:vector name="V" /> <!-- components depends on mesh dims -->
```

Example for tensors of a given rank, whose dimensionalities depend on the dimensionality $N$ of their assigned mesh:

```xml
<data:tensor rank="0" name="S" /> <!--     scalar -->
<data:tensor rank="1" name="V" /> <!--   N vector -->
<data:tensor rank="2" name="M" /> <!-- NxN matrix -->
<data:tensor rank="3" name="T" /> <!-- N^3 triad -->
```

Example for 5D user-defined data, which doesn't depend on the dimensionality of the mesh:

```xml
<data:custom size="5" name="V" /> <!-- always 5 components -->
```

A tensor of rank $r$ and dimensionality $N$ can be linearized to $\mathbb{R}^{N^r}$. Thus, the user-defined data dimensionality is the more general solution as it allows for odd configurations like 5D data.

**Implementation**

This can all be integrated into preCICE in a relatively straight-forward way. Mappings already map dimension-by-dimension.
VTK exporters need an update and may not be able to support all data types.

The API will need documentation for the general case of $N$ components, currently we use 2D and 3D examples.

Open question is how the gradient mapping works in such cases. What needs to be updated here.

**Describe alternatives you've considered**

`data:tensor` could be skipped as user-defined data superseeds it. However, depending on the user, defining the rank of the tensor may be more natural.

**Additional context**

The user-defined data synergies well with #1669 to making pseudo-meshes easier to use, which is a good replacement for #202.

This is a wanted feature for passing data in multi-scale scenarios. @IshaanDesai

**Comment by uekerman** [comment]
> The current workaround is to use either pseudo meshes or multiple meshes to communicate the data via preCICE, which is both tedious and inefficient.

Isn't the current workaround to define multiple data fields on the same mesh, e.g. the two-scale heat conduction tutorial:

https://github.com/precice/tutorials/blob/develop/two-scale-heat-conduction/precice-config.xml

```xml
  <data:scalar name="k_00" />
  <data:scalar name="k_01" />
  <data:scalar name="k_10" />
  <data:scalar name="k_11" />

<mesh name="macro-mesh" dimensions="2">
    <use-data name="k_00" />
    <use-data name="k_01" />
    <use-data name="k_10" />
    <use-data name="k_11" />
</mesh>
```

An easier intermediate solution would be to only add support for rank-2 tensors:

```xml
<data:tensor name="k" />

<mesh name="macro-mesh" dimensions="2">
    <use-data name="k" />
</mesh>
```

I don't see a direct connection to #202 and #1669. Am I missing anything?

**Comment by fsimonis** [comment]
> Isn't the current workaround to define multiple data fields on the same mesh, e.g. the two-scale heat conduction tutorial:

I got this mixed up with global data. Will adjust OP

> I don't see a direct connection to https://github.com/precice/precice/issues/202 and https://github.com/precice/precice/issues/1669. Am I missing anything?

#202: user-defined data allows exchanging data of any shape using a single vertex as global data, which can be more user-friendly than having to handle multiple `data` and vertices. 

#1669: 1D meshes are simpler to use as pseudo-meshes as they only contain the made-up id of the vertex. Currently with 2D meshes, the additional $y$ needs to be provided too.

**Comment by uekerman** [comment]
> https://github.com/precice/precice/issues/202: user-defined data allows exchanging data of any shape using a single vertex as global data, which can be more user-friendly than having to handle multiple data and vertices.

I am not sure. Tensors would still depend on the space location (i.e. the mesh). Global data does not depend on the space location.

> https://github.com/precice/precice/issues/1669: 1D meshes are simpler to use as pseudo-meshes as they only contain the made-up id of the vertex. Currently with 2D meshes, the additional y needs to be provided too.

I agree, but I still don't see the connection :confused: This issue is about "pseudo data", not "pseudo meshes", no?

**Comment by fsimonis** [comment]
"replace" is a too strong word. User-defined data synergies well with pseudo-meshes, which are easier to use once we get 1D meshes.

> This issue is about "pseudo data", not "pseudo meshes", no?

I wouldn't call it pseudo-data. Tensors are bound to the dimensionality of the mesh. User-defined data isn't.

**Comment by uekerman** [comment]
Mmh, I get your point to some extent, but not where we should head yet.
I would still favor simply adding support for rank-2 tensors as described [above](https://github.com/precice/precice/issues/2085#issuecomment-2397552739).

**Comment by kcooley-cvd** [comment]
I have a situation where I use preCICE to transfer production rates of several chemical species, for which this example from [OP](https://github.com/precice/precice/issues/2085#issue-2558539254) would be very useful:
```xml
<data:custom size="5" name="V" /> <!-- always 5 components -->
```
Currently each individual species gets its own `data:scalar` entry, which has to be uniquely named. This is doable but unnecessarily tedious for the person writing the config file. It also means a config file cannot be reused for a case with a different number of species, or even for one with the same number of species but of different names, even though the essential nature of the coupling has not changed. So ideally, `size` could be set by the adapter at runtime so that a single `precice-config.xml` can be reused.

**Comment by MakisH** [comment]
If I remember correctly (but only discussed directly), @j-signorelli was also interested in such a feature. Not because of having higher-order data, but to be able to exchange packs of independent scalars together, to save on function calls / memory accesses. @j-signorelli please elaborate, if this is still the case.

**Comment by j-signorelli** [comment]
@MakisH Yes that is correct I am still interested in this. We usually make sure our grids are the same between participants or we use "direct access to a received mesh", and so mapping is not an issue. In the context of the state variables for fluids (rho, rhoV, rhoE), the current "naive" approach to do it in preCICE would be:

```xml
<data:scalar name="rho"/>
<data:vector name="rhoV"/>
<data:scalar name="rhoE"/>
```
However, because this data in the codes I use is stored in a struct-of-arrays (SoA) organization for vectorization + GPU acceleration, right now I actually currently manually split the rhoV above into separate scalar components (rhoV1, rhoV2, rhoV3) to avoid re-arranging data after receiving it:

```xml
<data:scalar name="rho"/>
<data:scalar name="rhoV-1"/>
<data:scalar name="rhoV-2"/>
<data:scalar name="rhoV-3"/>
<data:scalar name="rhoE"/>
```
Because the number of state variables depends on the space dimension, it would be nice to have that feature mentioned by @kcooley-cvd to be able to set the dimension of the data in the code at runtime instead of manually needing to update the config when going from 2D --> 3D. It instead even simpler could just be something like,

```xml
<data:nscalars name="State"/>
```
But this wouldn't be necessary if SoA was an option (then the first config snippet would be fine).

However, the exact issue @kcooley-cvd has with chemical species has also been a pain point for some of my labmates as well that simulate chemically-reacting flows. @paulmartin996 can speak more on this if needed - he uses ~20 different data tags.

**Punchline: there is definitely interest within my group in both an SoA organization feature and a data type that allows packing of a set of independent scalars.**
